lib/action.ml

```diff
@@ -674,18 +674,26 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) (Buildkite_api :
                match FailedStepSet.is_empty failed_steps with
                | true -> Lwt.return []
                | false ->
-                 let%lwt slack_user_id =
+                 let%lwt slack_ids =
                    match mention_user_on_failed_builds cfg n with
-                   | false -> Lwt.return_none
+                   | false -> Lwt.return []
                    | true ->
-                     let email = extract_metadata_email n.build.meta_data |> Option.default "" in
-                     (match%lwt Slack_api.lookup_user ~ctx ~cfg ~email () with
-                     | Ok (res : Slack_t.lookup_user_res) -> Lwt.return_some res.user.id
-                     | Error e ->
-                       log#warn "couldn't match commit email %s to slack profile: %s" email e;
-                       Lwt.return_none)
+                     let to_slack_id email =
+                       match%lwt Slack_api.lookup_user ~ctx ~cfg ~email () with
+                       | Ok (res : Slack_t.lookup_user_res) -> Lwt.return_some res.user.id
+                       | Error e ->
+                         log#warn "couldn't match commit email %s to slack profile: %s" email e;
+                         Lwt.return_none
+                     in
+                     let author_emails =
+                       FailedStepSet.fold
+                         (fun (step : Buildkite_t.failed_step) acc ->
+                           if List.mem step.author acc then acc else step.author :: acc)
+                         failed_steps []
+                     in
+                     Lwt_list.map_s to_slack_id author_emails
                  in
-                 Lwt.return [ Slack.generate_failed_build_notification ?slack_user_id ~failed_steps n channel ])
+                 Lwt.return [ Slack.generate_failed_build_notification ~slack_ids ~cfg ~failed_steps n channel ])
```

Comment:
```suggestion
                 let slack_ids = List.filter_map Fun.id slack_ids in
                 Lwt.return [ Slack.generate_failed_build_notification ~slack_ids ~cfg ~failed_steps n channel ])
```

and then remove the filter in slack.ml

I think it makes more sense to filter before and have a simpler type for the `generate_failed_build_notification` function.